rewrite-java/src/main/java/org/openrewrite/java/RenameVariable.java

```diff
@@ -38,35 +45,95 @@ public RenameVariable(J.VariableDeclarations.NamedVariable variable, String toNa
         }
 
         @Override
-        public J.CompilationUnit visitCompilationUnit(J.CompilationUnit cu, P p) {
-            return super.visitCompilationUnit(cu, p);
+        public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, P p) {
+            // Variables declared in a class declaration are scoped to the class regardless of statement order.
+            List<Statement> declaringStatement = classDecl.getBody().getStatements().stream()
+                    .filter(it -> it instanceof J.VariableDeclarations)
+                    .filter(it -> ((J.VariableDeclarations)it).getVariables().stream().anyMatch(v -> v.equals(variable)))
+                    .collect(Collectors.toList());
+
+            if (!declaringStatement.isEmpty() && currentNameScope.size() == 0) {
+                currentNameScope.add(classDecl.getBody());
+            }
+            return super.visitClassDeclaration(classDecl, p);
         }
 
         @Override
         public J.Identifier visitIdentifier(J.Identifier ident, P p) {
-            if (!isConstant && ident.getSimpleName().equals(variable.getSimpleName()) &&
-                    isInSameNameScope(scope, getCursor()) &&
-                    !(getCursor().dropParentUntil(J.class::isInstance).getValue() instanceof J.FieldAccess)) {
+            if (ident.getSimpleName().equals(variable.getSimpleName())
+                    && isInSameNameScope(scope, getCursor())
+                    // The size of the stack will be 1 if the identifier is in the right scope.
+                    && currentNameScope.size() == 1
+                    && !(getCursor().dropParentUntil(J.class::isInstance).getValue() instanceof J.FieldAccess)) {
+
                 return ident.withName(toName);
             }
             return super.visitIdentifier(ident, p);
         }
 
         @Override
-        public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations multiVariable, P p) {
-            if (!isConstant && multiVariable.getVariables().contains(variable)) {
-                if (multiVariable.hasModifier(J.Modifier.Type.Static) && multiVariable.hasModifier(J.Modifier.Type.Final)) {
-                    isConstant = true;
+        public Statement visitStatement(Statement statement, P p) {
+            if (getCursor().getParent() != null
+                    && !(getCursor().getParent().getValue() instanceof J.CompilationUnit)
+                    && !(getCursor().getParent().getValue() instanceof J.ClassDeclaration)) {
+
+                // The current statement is out of scope of the previous variable declaration.
+                if (currentBlock != null && !getCursor().isScopeInPath(currentBlock)
+                        && currentNameScope.size() > 0 && !getCursor().isScopeInPath(currentNameScope.peek())) {
+                    currentNameScope.pop();
+                }
+            }
+            return super.visitStatement(statement, p);
+        }
```

Comment:
I find it a bit strange that scope is determined by finding a statement which isn't in the scope. 
In Java lexical scope is determined by what block a bit of code is in. I'm curious, why isn't the popping of frames from the current name scope done in `visitBlock`? I would have assumed that would be the natural place for it.

Reply:
I thought so too and tried that initially, but I couldn't find a way to cover certain types of cases. I may be misunderstanding something though.

```
public class A {
    int method() {
        for (control) {
            code ...
        }
        return N;
    }
}
```

In the example, the statements in the method are being processed and the for loop is finished. Any statements following the for loop need to be evaluated for a rename. A similar thing happens with inner classes, try/catches, etc.

I had code in the visitBlock, but it seemed redundant since it was repeated in the statements. So, I took it out.

Do you have any thoughts on how to handle those types of cases?

Reply:
In a `visitBlock`-based approach you could look at the parent of the block to find variables declared in for loops / try with resources / catch blocks / etc. I suspect that could end up cleaner since you'd no longer need any complex logic to detect when a block had ended.